datadog_checks_dev/datadog_checks/dev/tooling/license_headers.py

```diff
@@ -0,0 +1,116 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+import pathlib
+import re
+from collections import namedtuple
+from typing import Callable, Iterable, List, Optional
+
+from ..errors import SubprocessError
+from .constants import get_root
+from .git import git_show_file
+from .utils import get_license_header as get_default_license_header
+
+_COPYRIGHT_PATTERN = re.compile(
+    r"""
+    ^(?:\#\ \(C\)\s*.*?\s*)+\n      # Copyright holders
+    \#\ All\ rights\ reserved\s*\n
+    \#\ .*?(?i:license).*?$    # License info (matching string `license` while ignoring case)
+    """,
+    re.MULTILINE | re.VERBOSE,
+)
+
+LicenseHeaderError = namedtuple("LicenseHeaderError", ["message", "path"])
+
+
+def _get_previous(path):
+    """Returns contents of previous (origin/master) version of file at `path` if it exists, and `None` otherwise."""
+    # git_show_file relies on global context to compute the final path from a relative one,
+    # so we need to pass it the relative path it expects
+    relpath = path.relative_to(get_root())
+    try:
+        return git_show_file(str(relpath), "origin/master")
+    except SubprocessError:
+        return None
+
+
+def validate_license_headers(
+    check_path: pathlib.Path,
+    extra_ignore: Optional[Iterable[pathlib.Path]] = None,
+    *,
+    get_previous: Callable[[pathlib.Path], Optional[str]] = _get_previous,
+) -> List[LicenseHeaderError]:
+    """
+    Validate license headers under `check_path` and return a list of validation errors.
+
+    Paths on `extra_ignore` will be ignored.
+
+    Assumptions regarding which files require license header validation:
+    - Only python (*.py) files need a license header
+    - Code under hidden folders (starting with `.`) are ignored
+    - Files under the following folders are not checked, as it's where integration
+      testing environments are typically defined, where licenses may be a bit more heterogeneous:
+      - tests/docker
+      - tests/compose
+    """
+    root = check_path
+    default_ignore = {pathlib.Path("tests/docker"), pathlib.Path("tests/compose")}
+    ignoreset = frozenset(default_ignore.union(extra_ignore or set()))
+
+    def walk_recursively(path):
+        for child in path.iterdir():
+            # For directories, keep recursing unless folder is blacklisted
+            if child.is_dir():
+                # Skip hidden folders
+                if child.relative_to(child.parent).as_posix().startswith('.'):
+                    continue
+
+                # Skip blacklisted folders
+                relpath = child.relative_to(root)
+                if relpath in ignoreset:
+                    continue
+
+                yield from walk_recursively(child)
+                continue
+
```

Comment:
We should also skip any files in the `.gitignore` (ie `build/`, `dist/`)

Reply:
I'm reluctant to implement this, on the grounds that it adds too much complexity for the little value it actually seems to add. For example, I can't find any `build` or `dist` folders on my working copies of the repo (other than those under `.tox`, but those are being ignored because hidden folders are ignored).

Reply:
Locally I had these directories which produced a lot of unnecessary errors. I think it would be ideal for the command to work the same locally and on CI. How much extra work do you think it would be to add this logic?

Reply:
Alright, I'm working on supporting that, but I'm thinking that it might be better go on a separate PR.

Reply:
https://github.com/DataDog/integrations-core/pull/13439